pkgs/io_file/lib/src/vm_windows_file_system.dart

```diff
@@ -80,4 +84,110 @@ base class WindowsFileSystem extends FileSystem {
+    // call before the value is needed.
+    win32.GetLastError();
+
+    final f = win32.CreateFile(
+      path.toNativeUtf16(),
+      win32.GENERIC_READ | win32.FILE_SHARE_READ,
+      win32.FILE_SHARE_READ | win32.FILE_SHARE_WRITE,
+      nullptr,
+      win32.OPEN_EXISTING,
+      win32.FILE_ATTRIBUTE_NORMAL,
+      0,
+    );
+    if (f == win32.INVALID_HANDLE_VALUE) {
+      final errorCode = win32.GetLastError();
+      throw _getError(errorCode, 'open failed', path);
+    }
+    try {
+      // The result of `GetFileSize` is not defined for non-seeking devices
+      // such as pipes.
+      if (win32.GetFileType(f) == win32.FILE_TYPE_DISK) {
+        final highFileSize = arena<win32.DWORD>();
+        final lowFileSize = win32.GetFileSize(f, highFileSize);
+        if (lowFileSize == 0xffffffff) {
+          // Indicates an error OR that the low order word of the is actually
+          // that constant.
+          final errorCode = win32.GetLastError();
+          if (errorCode != win32.ERROR_SUCCESS) {
+            return _readUnknownLength(path, f);
+          }
+        }
+        final fileSize = highFileSize.value << 32 | lowFileSize;
+        if (fileSize == 0) {
+          return Uint8List(0);
+        } else {
+          return _readKnownLength(path, f, fileSize);
+        }
+      }
+      return _readUnknownLength(path, f);
+    } finally {
+      win32.CloseHandle(f);
+    }
+  });
+
+  Uint8List _readUnknownLength(String path, int file) => ffi.using((arena) {
+    final buffer = arena<Uint8>(blockSize);
+    final bytesRead = arena<win32.DWORD>();
+    final builder = BytesBuilder(copy: true);
+
+    while (true) {
+      if (win32.ReadFile(file, buffer, blockSize, bytesRead, nullptr) ==
+          win32.FALSE) {
+        final errorCode = win32.GetLastError();
+        // On Windows, reading from a pipe that is closed by the writer results
+        // in ERROR_BROKEN_PIPE.
+        if (errorCode == win32.ERROR_BROKEN_PIPE ||
+            errorCode == win32.ERROR_SUCCESS) {
+          return builder.takeBytes();
+        }
+        throw _getError(errorCode, 'read failed', path);
+      }
+
+      if (bytesRead.value == 0) {
+        return builder.takeBytes();
+      } else {
+        final typed = buffer.asTypedList(bytesRead.value);
+        builder.add(typed);
+      }
+    }
+  });
+
+  Uint8List _readKnownLength(String path, int file, int length) =>
+      ffi.using((arena) {
+        final buffer = ffi.malloc<Uint8>(length);
+        final bytesRead = arena<win32.DWORD>();
+        var bufferOffset = 0;
+
+        while (bufferOffset < length) {
+          if (win32.ReadFile(
+                file,
+                (buffer + bufferOffset).cast(),
+                min(length - bufferOffset, maxReadSize),
+                bytesRead,
+                nullptr,
+              ) ==
+              win32.FALSE) {
+            ffi.malloc.free(buffer);
```

Comment:
does arena-based allocation makes this redundant?

Reply:
I'm not using the arena for `buffer` because, unless there is an error, I return it on line 183.